File: codeforces/395/a.cpp
#include <bits/stdc++.h>

using namespace std;

const int N = 1e5 + 5;
int p[N], n, c[N];
pair< int, int > e[N];
vector< pair< int, int > > adj[N];

int find(int x) {
  return (p[x] < 0) ? x : p[x] = find(p[x]);
}
int cnt[N];
bool merge(int u, int v) {
  if ((u = find(u)) == (v = find(v)))
    return 0;
  if (p[u] < p[v])
    swap(u, v);
  p[v] += p[u];
  p[u] = v;
  return 1;
}
int main() {
  scanf("%d", &n);
  memset(p, -1, sizeof p);
  for (int i = 0; i < n-1; i++) {
    int u, v;
    scanf("%d %d", &u, &v);
    u--; v--;
    e[i] = make_pair(u, v);
  }
  for (int i = 0; i < n; i++)
    scanf("%d", c+i);
  int g = n;
  for (int i = 0; i < n-1; i++) {
    int u = e[i].first, v = e[i].second;
    if (c[u] == c[v])
      g -= merge(u, v);
  }
  memset(cnt, 0, sizeof cnt);
  for (int i = 0; i < n-1; i++) {
    int u = e[i].first, v = e[i].second;
    if (c[u] != c[v])
      cnt[u]++, cnt[v]++;
  }
  int ans = -1;
  for (int i = 0; i < n; i++)
    if (cnt[i] >= g-1)
      ans = i;
  if (ans != -1)
    printf("YES\n%d\n", ans+1);
  else
    puts("NO");

  return 0;
}

File: codeforces/395/a.py
n = int(input())
g = [[] for _ in range(n)]
for _ in range(n-1):
    u, v = map(int, input().split())
    u -= 1
    v -= 1
    g[u].append(v)
    g[v].append(u)
c = list(map(int, input().split()))
dif = [0 for _ in range(n)]
cnt = 0
for v in range(n):
    dif[v] = 0
    for u in g[v]:
        if c[u] != c[v]:
            dif[v] += 1
    cnt += dif[v] > 0
ans = -1
for v in range(n):
    if dif[v] + 1 >= cnt:
        ans = v+1
if ans >= 0:
    print('YES\n%d' % ans)
else:
    print('NO')